src/Avalonia.Interactivity/Interactive.cs

```diff
@@ -137,112 +120,117 @@ public void RemoveHandler<TEventArgs>(RoutedEvent<TEventArgs> routedEvent, Event
+        protected EventRoute BuildEventRoute(RoutedEvent e)
         {
-            Contract.Requires<ArgumentNullException>(e != null);
-
-            e.RoutedEvent.InvokeRaised(this, e);
+            e = e ?? throw new ArgumentNullException(nameof(e));
 
-            List<EventSubscription> subscriptions = null;
+            var result = new EventRoute(e);
+            var hasClassHandlers = e.HasRaisedSubscriptions;
 
-            if (_eventHandlers?.TryGetValue(e.RoutedEvent, out subscriptions) == true)
+            if (e.RoutingStrategies.HasFlagCustom(RoutingStrategies.Bubble) ||
+                e.RoutingStrategies.HasFlagCustom(RoutingStrategies.Tunnel))
             {
-                foreach (var sub in subscriptions.ToList())
-                {
-                    bool correctRoute = (e.Route & sub.Routes) != 0;
-                    bool notFinished = !e.Handled || sub.AlsoIfHandled;
+                IInteractive? element = this;
 
-                    if (correctRoute && notFinished)
+                while (element != null)
+                {
+                    if (hasClassHandlers)
                     {
-                        if (sub.InvokeAdapter != null)
-                        {
-                            sub.InvokeAdapter(sub.Handler, this, e);
-                        }
-                        else
-                        {
-                            sub.Handler.DynamicInvoke(this, e);
-                        }
+                        result.AddClassHandler(element);
                     }
+
+                    element.AddToEventRoute(e, result);
+                    element = element.InteractiveParent;
+                }
+            }
+            else
+            {
+                if (hasClassHandlers)
+                {
+                    result.AddClassHandler(this);
                 }
+
+                ((IInteractive)this).AddToEventRoute(e, result);
             }
+
+            return result;
         }
 
-        private List<EventSubscription> GetEventSubscriptions(RoutedEvent routedEvent)
+        private IDisposable AddEventSubscription(RoutedEvent routedEvent, EventSubscription subscription)
         {
-            if (!EventHandlers.TryGetValue(routedEvent, out var subscriptions))
+            _eventHandlers ??= new Dictionary<RoutedEvent, List<EventSubscription>>();
+
+            if (!_eventHandlers.TryGetValue(routedEvent, out var subscriptions))
             {
                 subscriptions = new List<EventSubscription>();
-                EventHandlers.Add(routedEvent, subscriptions);
+                _eventHandlers.Add(routedEvent, subscriptions);
             }
 
-            return subscriptions;
+            subscriptions.Add(subscription);
+
+            return new UnsubscribeDisposable(subscriptions, subscription);
         }
 
-        private IDisposable AddEventSubscription(RoutedEvent routedEvent, EventSubscription subscription)
+        private sealed class EventSubscription
```

Comment:
@MarchingCube do you think it is worth making this a `readonly struct`?

Reply:
I think so. I remember seeing this code but I figured that it might be slower since we were copying whole list when raising events. But since now we don't do that - `readonly struct` will surely reduce our allocations.